dapps/shop/src/pages/admin/settings/Console.js

```diff
@@ -1,15 +1,159 @@
 
+const abi = [
+  'function updateListing(uint256, bytes32, uint256)',
+  'event ListingCreated (address indexed party, uint indexed listingID, bytes32 ipfsHash)',
+  'event ListingUpdated (address indexed party, uint indexed listingID, bytes32 ipfsHash)'
+]
+
+/**
+ * Gets a marketplace listing's most recent IPFS hash by querying the ethereum network
+ * for relevant logs.
+ * Note: In the future we may want to consider storing the listing's most recent
+ * hash in the shop back-end DB.
+ *
+ * @param {Object} provider: ether.js provider.
+ * @param {Object} contract: ether.js contract object for the marketplace.
+ * @param {Integer} contractEpoch: block number at which contract was created.
+ * @param {string} seller: ethereum address of the seller
+ * @param {Integer} listingId: id of the listing in the marketplace contract.
+ * @returns {Promise<string>} IPFS hash
+ * @private
+ */
+async function _getListingLatestIpfsHash(
+  provider,
+  contract,
+  contractEpoch,
+  seller,
+  listingId
+) {
+  // Create a filter to get the ListingCreated events for the listing.
+  const listingCreatedFilter = contract.filters.ListingCreated(
+    seller,
+    listingId,
+    null
+  )
+  listingCreatedFilter.fromBlock = contractEpoch || 0
+
+  // Create a filter to get the ListingUpdated events for the listing.
+  const listingUpdatedFilter = contract.filters.ListingCreated(
+    seller,
+    listingId,
+    null
+  )
+  listingUpdatedFilter.fromBlock = contractEpoch || 0
+
+  // Get all the ListingCreated|Updated events and keep only the most recent.
+  const listingCreatedEvents = await provider.getLogs(listingCreatedFilter)
+  const listingUpdatedEvents = await provider.getLogs(listingUpdatedFilter)
+  let latestEvent
+  if (listingUpdatedEvents && listingUpdatedEvents.length > 0) {
+    latestEvent = listingUpdatedEvents[listingUpdatedEvents.length - 1]
+  } else if (listingCreatedEvents && listingCreatedEvents.length === 1) {
+    latestEvent = listingCreatedEvents[0]
+  } else {
+    throw new Error(
+      `Failed getting the listing's latest IPFS hash from contract's events`
+    )
+  }
+
+  // The IPFS hash for the listing is encoded as bytes32 in the logs data field.
+  const bytes32Hash = latestEvent.data
+  return getIpfsHashFromBytes32(bytes32Hash)
+}
+
+/**
+ * Update a marketplace listing's IPFS data by adding a field "shopIpfsHash" that points to
+ * the IPFS hash of the DShop associated with the listing.
+ *
+ * @param {Object} config: shop config
+ * @param {string} shopIpfsHash: IPFS hash for the DShop associated with the listing.
+ * @returns {Promise<void>}
+ */
+async function updateListing(config, shopIpfsHash) {
+  if (!shopIpfsHash) {
+    console.error('shop IPFS hash not provided')
+    return
+  }
+  if (shopIpfsHash.length !== 46) {
+    console.error('Invalid IPFS hash')
+    return
+  }
+
+  const enabled = await window.ethereum.enable()
+  if (!enabled) {
+    console.error('No web3 provider detected in the browser')
```

Comment:
I don't think this `enable()` API is universal.  web3.js doesn't support it and it's not in the Ethereum "standard", for instance.  Looks like it's probably part of EIP-2255, which is still draft, whatever that's worth.

Reply:
I was just blatantly copying the same pattern as existing functionality for creating a listing. From this [code](https://github.com/OriginProtocol/origin/blob/master/dapps/shop/src/pages/super-admin/shops/CreateListing.js#L32)

Do you know what's the "standard" or better way to do this?

Reply:
I don't know what the current defacto standard would be no.  I'm out of practice on multi-client web3 support.